libsplinter/src/network/peer_manager/peer_map.rs

```diff
@@ -0,0 +1,162 @@
+// limitations under the License.
+use std::collections::HashMap;
+
+use super::error::PeerUpdateError;
+
+#[derive(Clone)]
+pub struct PeerMetadata {
+    pub id: String,
+    pub endpoints: Vec<String>,
+    pub active_endpoint: String,
+}
+
+pub struct PeerMap {
+    peers: HashMap<String, PeerMetadata>,
+    redirects: HashMap<String, String>,
+    // Endpoint to peer id
+    endpoints: HashMap<String, String>,
+    // peer id to number of reconnection attempts
+    disconnected: HashMap<String, u64>,
+}
+
+/// A map of Peer IDs to peer metadata, which also maintains a redirect table for updated peer ids.
+///
+/// Peer metadata includes the peer_id, the list of endpoints and the current active endpoint.
+impl PeerMap {
+    pub fn new() -> Self {
+        PeerMap {
+            peers: HashMap::new(),
+            redirects: HashMap::new(),
+            endpoints: HashMap::new(),
+            disconnected: HashMap::new(),
+        }
+    }
+
+    /// Returns the current list of peer ids.
+    ///
+    /// This list does not include any of the redirected peer ids.
+    pub fn peer_ids(&self) -> Vec<String> {
+        self.peers
+            .iter()
+            .map(|(_, metadata)| metadata.id.to_string())
+            .collect()
+    }
+
+    pub fn peer_metadata(&self, peer_id: &str) -> Option<&PeerMetadata> {
+        self.peers.get(peer_id)
+    }
+
+    /// Insert a new peer id and endpoints
+    pub fn insert(&mut self, peer_id: String, endpoints: Vec<String>, active_endpoint: String) {
+        let peer_metadata = PeerMetadata {
+            id: peer_id.clone(),
+            endpoints: endpoints.clone(),
+            active_endpoint,
+        };
+
+        self.peers.insert(peer_id.clone(), peer_metadata);
+
+        for endpoint in endpoints {
+            self.endpoints.insert(endpoint, peer_id.clone());
+        }
+    }
+
+    /// Remove a peer id, its endpoint and all of its redirects. Returns the active_endpoint of
+    /// the peer.
+    pub fn remove(&mut self, peer_id: &str) -> Option<String> {
+        debug!("Removing peer: {}", peer_id);
+        self.redirects
+            .retain(|_, target_peer_id| target_peer_id != peer_id);
+        if let Some(peer_metadata) = self.peers.remove(&peer_id.to_string()) {
+            for endpoint in peer_metadata.endpoints.iter() {
+                self.endpoints.remove(endpoint);
+            }
+
+            Some(peer_metadata.active_endpoint)
+        } else {
+            None
+        }
+    }
+
+    /// Updates a peer id, and creates a redirect for the old id to the given new one.
+    ///
+    /// Additionally, it updates all of the old redirects to point to the given new one.
+    pub fn update_peer_id(
+        &mut self,
+        old_peer_id: String,
+        new_peer_id: String,
+    ) -> Result<(), PeerUpdateError> {
+        if let Some(peer_metadata) = self.peers.remove(&old_peer_id) {
+            let mut new_peer_metadata = peer_metadata.clone();
```

Comment:
Is this clone necessary? Can you just declare `peer_metadata` as `mut` above? Same for the `update_active_endpoint`.

Reply:
updated